src/Geta.NotFoundHandler.Optimizely/Infrastructure/Configuration/ServiceCollectionExtensions.cs

```diff
@@ -0,0 +1,81 @@
+
+namespace Geta.NotFoundHandler.Optimizely.Infrastructure.Configuration
+{
+    public static class ServiceCollectionExtensions
+    {
+        public static IServiceCollection AddOptimizelyNotFoundHandler(this IServiceCollection services)
+        {
+            return AddOptimizelyNotFoundHandler(services, _ => { });
+        }
+
+        public static IServiceCollection AddOptimizelyNotFoundHandler(
+            this IServiceCollection services,
+            Action<OptimizelyNotFoundHandlerOptions> setupAction)
+        {
+            services.AddSingleton<OptimizelySyncEvents>();
+            services.AddSingleton<ContentUrlHistoryEvents>();
+            services.AddTransient<Upgrader>();
+            services.AddSingleton<ContentLinkLoader>();
+            services.AddSingleton<ContentKeyGenerator>();
+            services.AddSingleton<ContentUrlLoader>();
+            services.AddSingleton<SqlContentUrlHistoryRepository>();
+            services.AddSingleton<IRepository<ContentUrlHistory>, SqlContentUrlHistoryRepository>();
+            services.AddSingleton<IContentUrlHistoryLoader, SqlContentUrlHistoryRepository>();
+            services.AddSingleton<ContentUrlIndexer>();
+            services.AddSingleton<RedirectBuilder>();
+
+            // Background service
+            services.AddSingleton<IAutomaticRedirectsService, DefaultAutomaticRedirectsService>();
+            services.AddSingleton<ChannelMovedContentRegistratorQueue>();
+            services.AddSingleton<IMovedContentRegistratorQueue, ChannelMovedContentRegistratorQueue>();
+            services.AddHostedService<MovedContentRegistratorBackgroundService>();
+
+            var providerOptions = new OptimizelyNotFoundHandlerOptions();
+            setupAction(providerOptions);
+            foreach (var provider in providerOptions.ContentKeyProviders)
+            {
+                services.AddSingleton(typeof(IContentKeyProvider), provider);
+            }
+
+            foreach (var provider in providerOptions.ContentLinkProviders)
+            {
+                services.AddSingleton(typeof(IContentLinkProvider), provider);
+            }
+
+            foreach (var provider in providerOptions.ContentUrlProviders)
+            {
+                services.AddSingleton(typeof(IContentUrlProvider), provider);
+            }
+
+            services.Configure<ProtectedModuleOptions>(
+                pm =>
+                {
+                    if (!pm.Items.Any(i => i.Name.Equals(Constants.ModuleName, StringComparison.OrdinalIgnoreCase)))
+                    {
+                        pm.Items.Add(new ModuleDetails { Name = Constants.ModuleName });
+                    }
+                });
+
+            services.AddOptions<OptimizelyNotFoundHandlerOptions>().Configure<IConfiguration>((options, configuration) =>
+            {
+                setupAction(options);
+                configuration.GetSection("Geta:NotFoundHandler:Optimizely").Bind(options);
```

Comment:
was wondering also in GoogleFeed - isn't it so that custom setup action is called by passing options for configure and then options are re-binded from config file?

idea: code and then config file (to overwrite something that could be set in file)?

Reply:
Yes, correct.
So that it overrides configuration in order - Default => Code => Config file
This allows us to use configuration files to configure settings differently in different environments without changing the code.